src/architecture/merkledb.md

```diff
@@ -0,0 +1,348 @@
+# Exonum MerkleDB
+
+Exonum MerkleDB is a persistent storage implementation based on
+[RocksDB][rocks-db]. MerkleDB provides APIs to work with merkelized data
+structures.
+
+MerkleDB is an object database. Objects represent the highest abstraction level
+for the [data storage](storage.md). All objects fall into the following groups:
+
+- **blobs** which represent sequences of bytes
+- **root objects** that do not have parents. These objects have UTF-8
+  identifiers, for example, "block", "state". Root objects can contain blob
+  items inside them.
+
+1. [Root Objects Types](#root-object-types) lists the types of data
+   stores supported by MerkleDB. Objects represent the highest abstraction
+   level for the data storage.
+2. [Low-level storage](#low-level-storage) explains how objects are persisted
+   using RocksDB.
+3. [View layer](#view-layer) describes the wrapper over the DB engine
+   that ensures atomicity of blocks and transactions.
+4. [Indexing](#indexing) explains how indices over structured data can be
+   built in MerkleDB.
+
+## Root Object Types
+
+Root objects perform the same role as tables in relational database
+management systems (RDBMSs). All MerkleDB objects internally are
+implemented as wrappers around key-value stores.
+
+Root objects in MerkleDB fall into several types:
+
+- **list** of items. Each item is preceded by an index which represents a `u64`
+  integer. Indices are implicitly defined by the items order
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents an optional single item.
+
+Both keys and values in the wrapped stores are persisted as byte sequences.
+Values are serialized according to the Protobuf serialization format.
```

Comment:
Nit: this is opt-in and (for better or worse) is not enforced on the MerkleDB level. For example, some core objects store hashes as values, and they are not stored as Protobuf's `bytes` or messages with a single `bytes` field, but rather directly as byte sequences.

Reply:
deleted mentioning of Protobuf serialization